### PR TITLE
password and willMessage field types should be byte[] in codec.mqtt

### LINKED ISSUE #6750: codec.mqtt: wrong password and willMessage field types in MqttConnectPayload/MqttEncoder (should be byte[])
In the present implementation of MqttConnectPayload/MqttEncoder, willMessage and password fields are  strings. However, spec (3.1.3.5) clearly mandates that the password filed should support arbitrary binary data:

> The Password field contains 0 to 65535 bytes of binary data prefixed with a two byte length field which indicates the number of bytes used by the binary data (it does not include the two bytes taken up by the length field itself). 

Same for willMessage (3.1.3.3):

> This field consists of a two byte length followed by the payload for the Will Message expressed as a sequence of zero or more bytes.

Thus, the relevant field and encoder bits must treat the password and willMessage fields as byte[] or ByteBuf (the former is probably enough).

This has important practical implications and the fix is straightforward and simple.

### PR DESCRIPTION
**Pull request by sbcd90**
Motivation:

Update the mqtt-codec based on mqtt spec (3.1.3.5).

Modification:

- Changes made to the file `MqttConnectPayload.java`.
- Subsequent changes have been made to files `MqttDecoder.java`, `MqttEncoder.java`, `MqttMessageBuilders.java`.
- Test cases have been updated.

Result:

Fixes #6750 .

### CODE REVIEW

**Comment by Scottmitch** [comment]
This is a duplicate of https://github.com/netty/netty/pull/6769 ... fyi @oakad if you don't plan on resolving the comment in your PR will will focus on this PR instead.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java [hunk 1]
```diff
@@ -50,15 +50,15 @@ public String willTopic() {
         return willTopic;
     }
 
-    public String willMessage() {
+    public byte[] willMessage() {
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java]: Please use the same approach (add a new method) as in https://github.com/netty/netty/pull/6769 and avoid breaking the public API.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java [hunk 2]
```diff
@@ -445,6 +438,21 @@ private static MqttFixedHeader decodeFixedHeader(ByteBuf buffer) {
         return new Result<String>(s, numberOfBytesConsumed);
     }
 
+    private static Result<byte[]> decodeByteArray(ByteBuf buffer, int minBytes, int maxBytes) {
+        final Result<Integer> decodedSize = decodeMsbLsb(buffer);
+        int size = decodedSize.value;
+        int numberOfBytesConsumed = decodedSize.numberOfBytesConsumed;
+        if (size < minBytes || size > maxBytes) {
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: can we remove `minBytes` and `maxBytes` parameters? `maxBytes` is always `MAX_VALUE` and `minBytes` is always `0`. can we instead just throw an exception (or return early) if `size < 0`?

**Comment by sbcd90** [comment]
Hi @Scottmitch , I have made the suggested changes.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java [hunk 3]
```diff
@@ -25,16 +26,30 @@
 
     private final String clientIdentifier;
     private final String willTopic;
-    private final String willMessage;
+    private final byte[] willMessage;
     private final String userName;
-    private final String password;
+    private final byte[] password;
 
     public MqttConnectPayload(
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java]: should we deprecate this constructor in favor of the new constructor which takes `byte[]`?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java [hunk 4]
```diff
@@ -51,6 +66,10 @@ public String willTopic() {
     }
 
     public String willMessage() {
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java]: Please deprecate this method and refer to the `willMessageInBytes` in the `@deprecated` javadocs.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java [hunk 5]
```diff
@@ -59,6 +78,10 @@ public String userName() {
     }
 
     public String password() {
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttConnectPayload.java]: Please deprecate this method and refer to the `passwordInBytes ` in the `@deprecated` javadocs.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java [hunk 6]
```diff
@@ -445,6 +434,21 @@ private static MqttFixedHeader decodeFixedHeader(ByteBuf buffer) {
         return new Result<String>(s, numberOfBytesConsumed);
     }
 
+    private static Result<byte[]> decodeByteArray(ByteBuf buffer) {
+        final Result<Integer> decodedSize = decodeMsbLsb(buffer);
+        int size = decodedSize.value;
+        int numberOfBytesConsumed = decodedSize.numberOfBytesConsumed;
+        if (size < 0) {
+            buffer.skipBytes(size);
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: this is invalid and will throw an exception.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java [hunk 7]
```diff
@@ -445,6 +434,21 @@ private static MqttFixedHeader decodeFixedHeader(ByteBuf buffer) {
...
+            numberOfBytesConsumed += size;
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: do you really want to decrement `numberOfBytesConsumed`?

Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: can you add a unit test which hits this code?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 8]
```diff
@@ -120,6 +121,11 @@ public ConnectBuilder willTopic(String willTopic) {
         }
 
         public ConnectBuilder willMessage(String willMessage) {
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: deprecate this method and reference the new method in javadocs?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 9]
```diff
@@ -146,6 +152,11 @@ public ConnectBuilder username(String username) {
         }
 
         public ConnectBuilder password(String password) {
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: deprecate this method and reference the new method in javadocs?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java (hunk 7 above)
Comment (sbcd90) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: This part of the code will be hit only in a hypothetical situation when the encoded bytebuf have some issues like length of a message field, say password is set to -1 etc.

However, if the assumption is, the encoded bytebuf is created out of the MqttEncoder, then this part of the code can be safely commented out.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java [hunk 10]
```diff
@@ -17,6 +17,8 @@
 package io.netty.handler.codec.mqtt;
 
 import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
```
Comment (netkins) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: MINOR Remove this unused import 'io.netty.buffer.ByteBufAllocator'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java [hunk 11]
```diff
@@ -17,6 +17,8 @@
...
+import io.netty.buffer.EmptyByteBuf;
```
Comment (netkins) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: MINOR Remove this unused import 'io.netty.buffer.EmptyByteBuf'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java [hunk 12]
```diff
@@ -445,6 +434,20 @@ private static MqttFixedHeader decodeFixedHeader(ByteBuf buffer) {
         return new Result<String>(s, numberOfBytesConsumed);
     }
 
+    private static Result<byte[]> decodeByteArray(ByteBuf buffer) {
+        final Result<Integer> decodedSize = decodeMsbLsb(buffer);
+        int size = decodedSize.value;
+        int numberOfBytesConsumed = decodedSize.numberOfBytesConsumed;
+        if (size < 0) {
+            numberOfBytesConsumed += size;
```
Comment (Scottmitch) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: I still don't understand this line of code. 2 bytes have still been consumed either way even if the data is invalid. This value is ultimately set to `bytesRemainingInVariablePart` which may be used later during the decode process. You mentioned this was a [hypothetical situation](https://github.com/netty/netty/pull/6817#discussion_r120771580) but it is not clear what will happen if this situation does occur (if its possible at all). If this situation is possible please add a unit test, and if it isn't possible can we remove this line?

**Comment by sbcd90** [comment]
Hi @Scottmitch, removed that line of code.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java [hunk 13]
```diff
@@ -445,6 +434,16 @@ private static MqttFixedHeader decodeFixedHeader(ByteBuf buffer) {
...
     }
 
+    private static Result<byte[]> decodeByteArray(ByteBuf buffer) {
+        final Result<Integer> decodedSize = decodeMsbLsb(buffer);
+        int size = decodedSize.value;
+        int numberOfBytesConsumed = decodedSize.numberOfBytesConsumed;
+        byte[] bytes = new byte[size];
+        buffer.readBytes(bytes);
+        numberOfBytesConsumed += size;
+        return new Result<byte[]>(bytes, numberOfBytesConsumed);
```
Comment (fenik17) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java]: Now a `numberOfBytesConsumed` is not needed.
```java
return new Result<byte[]>(bytes, decodedSize.numberOfBytesConsumed + size);
```

**Comment by sbcd90** [comment]
Hi @fenik17 , @Scottmitch , do you think of any further changes?

**Comment by normanmaurer** [comment]
@sbcd90 can you please sign our ICLA and let me know once done:

https://netty.io/s/icla

**Comment by sbcd90** [comment]
@normanmaurer , I have signed the ICLA.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (66c83f7b74b3ea28baa4a0702d621aba8331993a).

@sbcd90 thanks!

**Pull request closed without merging**